File: source/CSharp/HalfIntervalMethod.cs
using System;

namespace Exploring.Net.CSharp {
    public class HalfIntervalMethod {
        public HalfIntervalMethod(Func<double, double> function) {
            this.function = function;
        }

        public double FindRoot(Interval interval) {
            var values = interval.Map(function);
            if (!values.CrossesZero) {
                throw new ArgumentException("Values are not opposite signs.");
            }
            return FindRootByHalfIntervals(values.Ascending ? interval: interval.Swap);
        }

        double FindRootByHalfIntervals(Interval interval) {
            while (interval.Size >= 1E-10) {
                var value = function(interval.MidPoint);
                if (value == 0.0) break;
                interval = value < 0.0 ? interval.RightHalf : interval.LeftHalf;
            }
            return interval.MidPoint;
        }

        readonly Func<double, double> function;
    }
}

File: source/FSharp/HalfIntervalMethod.fs
namespace Exploring.Net.FSharp

open Interval

module HalfIntervalMethod = 

    let findRoot aFunction interval =

        let rec findRootByHalfInterval interval =
            if size interval < 1E-10
                then midpoint interval
                else
                    match aFunction (midpoint interval) with
                        | 0.0 -> midpoint interval
                        | value -> findRootByHalfInterval ((if value < 0.0 then rightHalf else leftHalf) interval)

        let values = map aFunction interval
        if crossesZero values 
            then findRootByHalfInterval (if ascending values then interval else swap interval)
            else raise (new System.ArgumentException("Values are not opposite signs."))

    let findSineRoot interval = findRoot sin interval
